engine/src/main/java/io/zeebe/engine/processing/bpmn/container/ProcessProcessor.java

```diff
@@ -62,56 +65,30 @@ public void onComplete(
 
     eventSubscriptionBehavior.unsubscribeFromEvents(context);
 
-    final var parentProcessInstanceKey = context.getParentProcessInstanceKey();
-    BpmnElementContext parentElementInstanceContext = null;
-    if (parentProcessInstanceKey > 0) {
-      // needs to be done before we delete the Process element instance
-      parentElementInstanceContext = stateBehavior.getParentElementInstanceContext(context);
-    }
+    final var onCompletedAction = getPostTransitionAction(element, context);
 
     // we need to send the result before we transition to completed, since the
     // event applier will delete the element instance
     processResultSenderBehavior.sendResult(context);
 
     final var completed = stateTransitionBehavior.transitionToCompleted(context);
-
-    if (parentElementInstanceContext != null) {
-      stateTransitionBehavior.onCalledProcessCompleted(completed, parentElementInstanceContext);
-    }
-
-    if (element.hasMessageStartEvent()) {
-      bufferedMessageStartEventBehavior.correlateMessage(completed);
-    }
+    onCompletedAction.accept(completed);
   }
 
   @Override
   public void onTerminate(
       final ExecutableFlowElementContainer element, final BpmnElementContext context) {
+
     eventSubscriptionBehavior.unsubscribeFromEvents(context);
+    incidentBehavior.resolveIncidents(context);
 
     final var noActiveChildInstances = stateTransitionBehavior.terminateChildInstances(context);
 
     if (noActiveChildInstances) {
-
-      final var parentProcessInstanceKey = context.getParentProcessInstanceKey();
-      BpmnElementContext parentElementInstanceContext = null;
-      if (parentProcessInstanceKey > 0) {
-        // needs to be done before we delete the Process element instance
-        parentElementInstanceContext = stateBehavior.getParentElementInstanceContext(context);
-      }
+      final var onTerminatedAction = getPostTransitionAction(element, context);
```

Comment:
One minor comment here is that this solution, in my opinion, is not very elegant.
Most uses of this method have a strong pattern in terms of follow up lines:
```
onTransitionAction = getPostTransitionAction(element, context)
transitioned = stateTransitionBehavior.transitionToX(context)
onTransitionAction.accept(transitioned)
``` 

Maybe the code would be clearer with a construct like `BpmnStateTransitionBehavior#invokeElementContainerIfPresent()`

So something like:
```
transitioned = transitionAndFollowUp(element, context, stateTransitionBehavior::transitionToX)
```